### PR TITLE
Fix multiplier issues with validation, and refactor validation logic

### PR DESCRIPTION
**Pull request by damian0815**
There's some changes in here to how multiplier.txt is applied. 

Mainly, the logic has been simplified and moved to runtime rather than load time. On `DataLoaderMultiAspect` multipliers are lazily applied in `__pick_multiplied_set()`, rather than being pre-baked into `prepared_train_data`. The main implication of this is that for fractional multipliers, EveryDream will now train with a different random subset of the folder each epoch, rather than picking the subset once on load and then re-using the same subset every epoch.

I have not extensively tested this - please check that it works as expected with fractional and also >1 multipliers before merging.

### CODE REVIEW

File: data/data_loader.py [hunk 1]
```diff
@@ -41,40 +42,27 @@ def __init__(self, image_train_items: list[ImageTrainItem], seed=555, batch_size
...
-                    picked_images.append(iti)
-                    remaining -= 1
-                    iti.multiplier = 0.0
-                if remaining <= 0:
-                    break
-        
-        del data_copy
+                multiplier -= 1
+            # deal with fractional remainder
+            if multiplier > randomizer.uniform(0, 1):
```
Comment (damian0815) [inline data/data_loader.py:63]: Hmm. Ok actually there's still a problem with the logic here. I don't have more time to fix this now, will do it later. Basically this needs to a) ensure the epoch size stays the same and b) pick a subset per-folder based on the folder's shared multiplier.

**Comment by damian0815** [comment]
Multipliers are now being applied dynamically, in a way that preserves a constant epoch size. 

Two implications of the current method in this PR:

#### 1.
Fractional multipliers are applied globally, not per-folder. If you have a folder of images `folderA` with multiplier 0.2 and another folder `folderB` with multiplier 0.8, there is no guarantee that you'll get 20% of `folderA` and 80% of `folderB`. Instead, after taking care of whole-number multipliers, any images from `folderA` will have a 20% chance of being selected to top up the epoch, while images from `folderB` will have an 80% chance.

#### 2.
When using fractional dropout the multiplier is completely ignored. I don't know what fractional dropout is for, but I suspect that the correct way to do this is to replace this if:
```
    if dropout_fraction < 1.0:
        picked_images = self.__pick_random_subset(dropout_fraction, randomizer)
    else:
        picked_images = self.__pick_multiplied_set(randomizer)
```
with something more like this:
```
    picked_images = self.__picked_multiplied_set(randomizer)
    if dropout_fraction < 1.0:
        picked_images = self.__apply_dropout(picked_images, dropout_fraction, randomizer)
```

**Comment by damian0815** [comment]
Fractional multipliers are now applied per-directory.

File: data/data_loader.py [hunk 2]
```diff
@@ -38,43 +39,40 @@ def __init__(self, image_train_items: list[ImageTrainItem], seed=555, batch_size
...
-                    iti.multiplier = 0.0
-                if remaining <= 0:
-                    break
-        
-        del data_copy
+                multiplier -= 1
+            # fractional remainders must be dealt with separately
+            if multiplier > 0:
+                directory = os.path.dirname(iti.pathname)
+                fractional_images_per_directory[directory].append(iti)
```
Comment (victorchall) [inline data/data_loader.py:67]: This is going to cause problems in the future if (when) multiplier is detached from directory.  I purposely made it a property attached to each image.  The multiplier coming from the directory is an artifact that shouldn't be promulgated throughout the codebase.

File: data/data_loader.py [hunk 3]
```diff
@@ -38,43 +39,40 @@ def __init__(self, image_train_items: list[ImageTrainItem], seed=555, batch_size
         self.prepared_train_data = image_train_items
         random.Random(self.seed).shuffle(self.prepared_train_data)
         self.prepared_train_data = sorted(self.prepared_train_data, key=lambda img: img.caption.rating())
+        expected_epoch_size = math.floor(sum([i.multiplier for i in self.prepared_train_data]))
```
Comment (victorchall) [inline data/data_loader.py:42]: This is a better way to do this than before

File: data/data_loader.py (hunk 2 above)
Comment (victorchall) [reply data/data_loader.py:67]: I don't think the shuffle/bifurcation is the best algorithm here.  Iteration is fast outside some corner cases like someone using a multiplier like 2.00000001 maybe.
